### PR TITLE
Coordinate Space Refactor + Separate Physics World Per Scene

### PR DESCRIPTION
**Pull request by HexaField**
## Summary

- [x] Pass physicsWorld into all physics functions
- [x] Rename SceneComponent “children” field to “scenes”
- [x] Separate physics world for each auto generated scene component
- [x] Change XR spaces
- [x] Scale scenes rather than camera

## References
closes #_insert number here_

### CODE REVIEW

File: packages/engine/src/scene/systems/ShadowSystem.tsx [hunk 1]
```diff
@@ -269,7 +269,14 @@ const EntityChildCSMReactor = (props: { rendererEntity: Entity }) => {
 
 function _CSMReactor() {
   const rendererEntity = useEntityContext()
-  const renderSettingsEntity = useChildWithComponent(rendererEntity, RenderSettingsComponent)
+  const renderer = useComponent(rendererEntity, RendererComponent).value
+  /**
+   * @todo Currently this will just return the first entity with a RenderSettingsComponent found,
+   *   but we need some more advanced rule for determining which entity to use
+   *   considering multi-scene support and spatial volumes.
+   * note: index 0 is local floor entity, so use index 1 (will usually be the first loaded scene)
```
Comment (heysokam) [inline packages/engine/src/scene/systems/ShadowSystem.tsx]: there is an `@note` tag in jsdoc, fyi :+1:

File: packages/engine/src/avatar/components/AvatarControllerComponent.ts [hunk 2]
```diff
@@ -101,6 +102,16 @@ export const AvatarControllerComponent = defineComponent({
     const avatarControllerComponent = useComponent(entity, AvatarControllerComponent)
     const isCameraAttachedToAvatar = useHookstate(getMutableState(XRControlsState).isCameraAttachedToAvatar)
     const camera = useComponent(Engine.instance.cameraEntity, CameraComponent)
+    const world = Physics.useWorld(entity)
+
+    useEffect(() => {
+      console.log('Creating character controller', world)
```
Comment (heysokam) [inline packages/engine/src/avatar/components/AvatarControllerComponent.ts]: pointing at this just in case it was missed :writing_hand:

File: packages/spatial/src/physics/classes/Physics.ts [hunk 3]
```diff
@@ -79,32 +86,135 @@ import {
...
+function createRigidBody(world: PhysicsWorld, entity: Entity) {
+  const sceneEntity = getAncestorWithComponent(entity, SceneComponent)
+
+  TransformComponent.getMatrixRelativeToEntity(entity, sceneEntity, mat4)
+  mat4.decompose(position, rotation, scale)
 
-function createRigidBody(entity: Entity, world: World) {
-  const transform = getComponent(entity, TransformComponent)
-  transform.matrixWorld.decompose(position, rotation, scale)
+  console.log('createRigidBody', scale.x, scale.y, scale.z)
```
Comment (heysokam) [inline packages/spatial/src/physics/classes/Physics.ts]: :point_up: just in case

**Review by MichaelEstes — COMMENTED** [review]
Is there any updating of scenes that needs to happen for this branch? CSM aren't working for me on this branch as is.

**Comment by HexaField** [comment]
> Is there any updating of scenes that needs to happen for this branch? CSM aren't working for me on this branch as is.

I think it's faced some entropy. I'll have to go QA and fix things next week.

**Comment by HexaField** [comment]
> Is there any updating of scenes that needs to happen for this branch? CSM aren't working for me on this branch as is.

Fixed

**Comment by MichaelEstes** [comment]
I can't load into sky-station without falling through the ground, every other location works though so I'm not sure if that's an issue with how it's set up. Also occasionally I can't move the avatar after loading into a location. Everything else looks good

**Comment by HexaField** [comment]
> I can't load into sky-station without falling through the ground, every other location works though so I'm not sure if that's an issue with how it's set up. Also occasionally I can't move the avatar after loading into a location. Everything else looks good

Urgh :(
Screenshot from 2024-07-24 09-28-39

**Comment by HexaField** [comment]
I have no idea what's going on. The transforms are all correct, but the rigidbodies themselves seem to be somehow not syncing? No idea how/when this regressed.

**Comment by HexaField** [comment]
> I can't load into sky-station without falling through the ground, every other location works though so I'm not sure if that's an issue with how it's set up. Also occasionally I can't move the avatar after loading into a location. Everything else looks good

Fixed. Came down to use not using the correct reference space in updating collider pose from transforms, and updating transforms from rigidbody [ef649bb](https://github.com/EtherealEngine/etherealengine/pull/10441/commits/ef649bb4f4a2e4055824ae64c051ba81344675f2)

File: packages/engine/src/avatar/functions/moveAvatar.ts [hunk 4]
```diff
@@ -413,7 +418,7 @@ const _updateLocalAvatarRotationAttachedMode = (entity: Entity) => {
   const rigidbodyForward = new Vector3(0, 0, -1).applyQuaternion(rigidbody.targetKinematicRotation).setY(0)
   const angle = viewerForward.angleTo(rigidbodyForward)
 
-  if (angle > Math.PI * 0.25 || rotationNeedsUpdate == true) {
+  if (angle > Math.PI / 12 || rotationNeedsUpdate == true) {
```
Comment (AidanCaruso) [inline packages/engine/src/avatar/functions/moveAvatar.ts:421]: Why was this changed from multiplication to division?

**Review by AidanCaruso — APPROVED** [review]
Looks good, tested colliders of multiple scenes loaded together along with webxr.

File: packages/engine/src/avatar/functions/moveAvatar.ts (hunk 4 above)
Comment (HexaField) [reply packages/engine/src/avatar/functions/moveAvatar.ts:421]: 0.25 was much too high a value, it was sometimes not updating. we can precompute the division to optimize but it's only happening once or twice per frame
